### PR TITLE
Pattern Library: Pre-populate react-query cache for individual categories after fetching all patterns

### PR DESCRIPTION
**Pull request by fredrikekelund**
Related to https://github.com/Automattic/dotcom-forge/issues/6532

## Proposed Changes

Data fetching for the pattern library works like this:
- Patterns are always fetched from the `/ptk/patterns` API endpoint
- When the user requests an individual category, we query the API endpoint for patterns from that category
- When the user submits a global search, we query the API endpoint for _all_ patterns (and then we filter them client-side)

> [!IMPORTANT]
> There's a gotcha to the `usePatterns` query for _all_ patterns, which is that we currently submit it instantly when the home page loads, we don't wait for the user to submit a search (although this is the conceptual intention). #89478 brings back the intended behavior.

After we've fetched all patterns, we don't technically need to run new queries to fetch patterns from individual categories. We already have all the data necessary to display those patterns.

This PR takes the first step to avoid those unnecessary requests by prepopulating the react-query cache for individual category queries with the data from the _all_ patterns request.

In other words, after the user has submitted a search, navigating to a new category will be instant – all patterns will already be loaded.

## Testing Instructions

1. Navigate to `/patterns`
2. Wait a second or two
3. Click on the `About` category
4. Ensure that the pattern previews are there instantly (there is no short moment when the list of patterns is empty)

### CODE REVIEW

File: client/my-sites/patterns/lib/filter-patterns-by-type.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { CATEGORY_PAGE } from 'calypso/my-sites/patterns/constants';
+import { PatternTypeFilter, Pattern } from 'calypso/my-sites/patterns/types';
+
+// Filter patterns by `PatternTypeFilter` type
```
Comment (nightnei) [inline client/my-sites/patterns/lib/filter-patterns-by-type.ts]:
```suggestion
```
IMO the function and props names are self explanatory

File: client/my-sites/patterns/hooks/use-patterns.ts [hunk 2]
```diff
@@ -66,7 +67,42 @@ export function getPatternsQueryOptions(
...
-	return useQuery< Pattern[] >( getPatternsQueryOptions( locale, category, queryOptions ) );
+	const queryClient = useQueryClient();
+	const queryResult = useQuery< Pattern[] >(
+		getPatternsQueryOptions( locale, category, queryOptions )
+	);
+
+	// When users submit a search, we fetch *all* patterns. This callback groups the returned
+	// patterns by category and populates the react-query cache for each category.
+	useEffect( () => {
+		if ( category === '' && queryResult.data?.length ) {
```
Comment (nightnei) [inline client/my-sites/patterns/hooks/use-patterns.ts:80]: `category === ''` means that we run it every time when a users visits home page. WDYT about removing `category` from dependencies of the hood and introduce something like `isFirstVisitOfHomePage`?

**Review by nightnei — APPROVED** [review]
LGTM and works as expected.
Left just 2 not critical comments

File: client/my-sites/patterns/hooks/use-patterns.ts (hunk 2 above)
Comment (fredrikekelund) [reply client/my-sites/patterns/hooks/use-patterns.ts:80]: Yeah, it means we run it every time the home page is rendered (if `queryResult.dataUpdatedAt` differs from the last render). I don't think that's really a problem, though. We only prepopulate the category queries if they don't already have cached data.
